### ISSUE TITLE
ImageGatherExtended does not appear to work

### ISSUE THREAD

**Issue by riperiperi** [KhronosGroup/MoltenVK#2034]
MoltenVK reports `shaderImageGatherExtended` as true, but using it seems to cause an error:

```
00:00:11.985 |E| .NET ThreadPool Worker [mvk-error] VK_ERROR_INITIALIZATION_FAILED: Shader library compile failed (Error code 3):
00:00:11.985 |E| .NET ThreadPool Worker program_source:334:48: error: no matching member function for call to 'gather'
00:00:11.985 |E| .NET ThreadPool Worker     float4 _68 = spvDescriptorSet2.fp_t_tcb_16.gather(spvDescriptorSet2.fp_t_tcb_16Smplr, float2(_63, _65), _1164, component::x).xyzw;
00:00:11.985 |E| .NET ThreadPool Worker                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~
00:00:11.985 |E| .NET ThreadPool Worker /System/Library/PrivateFrameworks/GPUCompiler.framework/Versions/32023/Libraries/lib/clang/32023.21/include/metal/metal_texture:5164:24: note: candidate function not viable: no known conversion from 'const constant spvUnsafeArray<int2, 4>' to 'metal::int2' (aka 'int2') for 3rd argument
00:00:11.985 |E| .NET ThreadPool Worker   METAL_FUNC vec<T, 4> gather(sampler s, float2 coord, int2 offset = int2(0), component c = component::x) const constant METAL_CONST_ARG(c)
...
```

Looking at the latest MSL spec, I don't see any overloads that handle providing a fixed array of offsets. I also don't see anything special in spirv-cross to emulate this.

This used to "work" in a previous version, but I don't know if it was just broken before or actually worked. The commit where things changed might be https://github.com/KhronosGroup/SPIRV-Cross/commit/1047c13d40a2fccd5198ad8992618af2dab0303f, since that's the only thing that changed around texture gather at that time.

**Comment by cdavis5e** [comment]
It was broken, because it wasn't passing down the offsets at all.

There is no support for this in Metal. While it would be easy to emulate by doing four `read()`s, the problem is handling sampler wrap modes, which from my reading of the Vulkan standard affect the operation of the `OpImageGather` SPIR-V instruction and, indeed, Metal's `gather()` method, but which don't affect `read()`. We do not have this information when the shader is compiled, and there's no way to get it from the sampler.

**Comment by riperiperi** [comment]
You can center the coordinate on a texel on a regular texture sample instead of using a texel fetch to still respect the sampler wrap mode. Regardless, this feature should be reported as false if it’s not supported.

**Comment by billhollings** [comment]
> Regardless, this feature should be reported as false if it’s not supported.

The reason that `shaderImageGatherExtended` is enabled, is that it covers two features, one of which is supported by Metal, and the flag is enabled to allow support for that.

Eventually, the other will be covered by the final version of the Metal Portability extension. See [this discussion](https://github.com/KhronosGroup/Vulkan-Portability/issues/31) for more info.

**Comment by rcaridade145** [comment]
@billhollings   https://github.com/KhronosGroup/SPIRV-Cross/pull/2325 Fixes this no? @V380-Ori does it fix Ryujinx?

**Comment by V380-Ori** [comment]
> [@billhollings](https://github.com/billhollings) [KhronosGroup/SPIRV-Cross#2325](https://github.com/KhronosGroup/SPIRV-Cross/pull/2325) Fixes this no? [@V380-Ori](https://github.com/V380-Ori) does it fix Ryujinx?

Yes, this issue can be absolutely closed; it works in Ryujinx, and there’s even a workaround which emulates this functionality in Ryujinx that talented riperiperi made back when this was not supported in MoltenVK/SPIR-V.

Interestingly, couple months back when I tested Ryujinx with newer MoltenVK, L.A. Noire had a minor graphical issue when it was not emulated, but now I don’t see that issue anymore… 

@rcaridade145 I will update that MR for Ryubing, thank you!

**Comment by rcaridade145** [comment]
@V380-Ori  I asked because on your code you still had && !isMoltenVK https://git.ryujinx.app/ryubing/ryujinx/-/merge_requests/142/diffs#3fc57f8e513adebe14bcbd4848e7c0a90924bf09_779_778
